chrome/content/zotero/itemTree.jsx

```diff
@@ -369,6 +369,8 @@ var ItemTree = class ItemTree extends LibraryTree {
 		if (type == 'search' && action == 'modify') {
 			// TODO: Only refresh on condition change (not currently available in extraData)
 			await this.refresh();
+			// Redraw the tree to remove any old rows that do not match the filter
+			this.tree.invalidate();
```

Comment:
I am still tracking down exactly the best way to tackle this issue I ran into and I don't think I have a 100% good solution yet but here are my findings.

1. This is a separate issue from the ordering discussed earlier.
2. The issue can be seen by taking any saved search with 2 conditions in "match any" mode and removing one of the conditions. Without clicking on anything else, the rows that match the removed condition will still stick around until you click something, and only then they go away.
3. Adding a new condition to saved search will (sometimes?) add rows but display them as context rows until you click around, in which case they stop being greyed outr. 
4. Adding a condition to the saved search that matches a child of an already matching collapsed container row does not expand it until you click around. Same happens the other way around when a child stops matching and the container needs to be collapsed.
5. Note that this only happens when there is not selection in itemTree - if any row in itemTree is selected, things works as expected.
6. Also sometimes this does not seem to happen, so there may be some kind of race condition involved.


As I see, part of the issue can be that the tree invalidation in `sort()` that `refresh` calls is never reached. In our case of removing a saved search condition, `sort` [returns](https://github.com/zotero/zotero/blob/436b19d64cb942644ea80705b5247dc52e6ebeb7/chrome/content/zotero/itemTree.jsx#L1243) before getting to the tree invalidation, so I think the tree is just never updated in that case.

I just pushed a commit that seems to partly address these things. Checking if `sort()` ran and invalidated the tree works to remove the non-matching rows on condition change but does not (always) help with displaying newly matching children

Reply:
From my experimentation so far, just keeping `this.tree.invalidate();` after  `await this.refresh();`  on `search` `modify` notification seems to solve all of the issues above, unlike invalidating the tree from inside of `this.refresh` as in my current last commit. I assume it has something to do with when the tree invalidation runs

Reply:
To address number 4 from above,  we could do something like
```
if (type == 'search' && action == 'modify') {
	// TODO: Only refresh on condition change (not currently available in extraData)
	await this.refresh();
	// Redraw the tree to remove any old rows that do not match the filter
        if (!this.selection.focused) {
            this.tree.invalidate();
        }
}
```
That way, it would only run when there is no selection, which is when these issues (most importantly number 4 that's no handled by adding manual tree invalidation in refresh) happen. But it's not a fantastic solution since it doesn't address the actual reason of why it happens.

Reply:
From logs, i see that there are calls to invalidate the tree but `render` in itemTree is not called the last time with all the rows added to it after refresh. For some reason, invalidating the tree a bit after `refresh` as I did earlier (or even manually in Run Javascript) does result in `itemTree.render` being called.  Where can I find the mechanism connecting the tree invalidation and call to `render` in itemTree? I thought one always follows the other but that's clearly not the case.

Also, I have a [brief video](https://www.dropbox.com/scl/fi/j4chxg19bkai527smv9v1/saved_search_rendering.mov?rlkey=eqbznsfq3emfml268nwk3ea0w&dl=0) of the issue, just to make sure it's clear what I'm looking at

Reply:
This issue does seem to boil down to the tree not being invalidated in a few very specific situations when no item is selected in itemTree of saved search: a) when items should only be removed after, say, a condition update that only matches as subset of current items, so nothing is done in `sort()` where tree invalidation would happen otherwise b) when a parent needs to be expanded to reveal a matching child, `_restoreSelection`(which would invalidate rows by toggling them) returns right away if there is no selected item. 

I re-pushed my earlier changes to only invalidate the tree in `refresh` in those specific situations and it seems to address this for me